service-mirror.go

```diff
@@ -34,110 +44,55 @@ func main() {
 	//Shows line number: Too long
 	// log.SetReportCaller(true)
 
-	log.Infof("Starting Global Mirror")
+	log.Info("Starting Global Mirror")
 
 	var kubeconfig *string
 	if home := homedir.HomeDir(); home != "" {
 		kubeconfig = flag.String("kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) absolute path to the kubeconfig file")
 	}
 
+	//Specify the NameSpace for install controller & global svc.
+	globalSvcNs := flag.String("globalsvc-ns", GLOBAL_SVC_NAMESPACE, "(optional) Namespace to install service mirror controller and global mirror services.")
+
 	flag.Parse()
 
 	// use the current context in kubeconfig
-	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	config, err := client.BuildConfigFromFlags("", *kubeconfig)
 	if err != nil {
-		log.Panicf("Probably running Inside Cluster :%s", err.Error())
+		log.Panicf("Probably running Inside Cluster :%v", err.Error())
 	}
 
 	// creates the clientset
 	client, err := kubernetes.NewForConfig(config)
 	if err != nil {
-		log.Panicf("Issue in building client from config : %s", err.Error())
+		log.Panicf("Issue in building client from config : %v", err.Error())
 	}
 
-	/* Get running services in all the names, labelled with "mirror.linkerd.io/mirrored-service: true"
-	and which service which does have label : mirror.linkerd.io/headless-mirror-svc-name. Build selector for it.
-	*/
-	mirrorSvcLabel, err := labels.NewRequirement("mirror.linkerd.io/mirrored-service", selection.Equals, []string{"true"})
-	if err != nil {
-		log.Errorf("Unable to generate error requirement : %s", err.Error())
-	}
-
-	// Create label requirements for "mirror.linkerd.io/headless-mirror-svc-name" not existing
-	mirrorSvcParentLabel, err := labels.NewRequirement("mirror.linkerd.io/headless-mirror-svc-name", selection.DoesNotExist, []string{})
-	if err != nil {
-		log.Errorf("Unable to generate error requirement : %s", err.Error())
+	//Create context
+	ctx := &Context{
```

Comment:
Okay, I actually don't think we should support this at all. It adds a bit of complexity (moar indirection!!!). Here's my view on it.

* we maintain a struct definition that wraps around a logger, a context and a client
* we only use this to pass it in the constructor
* service watcher then calls them by going through this composed struct `watcher.ctx.log` (imagine writing that for every single log line).

I would remove it personally. If you want it to stay in and you anticipate it will hold more information, we can embed it instead so the fields are flattened

```go
struct Watcher{
  Context
}

// and now we can do watcher.log
```

We can always bring it back in the future. Remember, it's easier to add something new in than it is to remove something.

Reply:
Yess this make sense let me refactor it and try to do in suggested way. Its odd rn.